boltons/ioutils.py

```diff
@@ -0,0 +1,322 @@
+    unicode.
+    """
+    if isinstance(string, text_type):
+        return string.encode('utf-8')
+    elif isinstance(string, binary_type):
+        return string
+    else:
+        raise TypeError("{} is not a unicode or str object".format(string))
+
+
+class SpooledIOBase(object):
+    """
+    The SpooledTempoaryFile class doesn't support a number of attributes and
+    methods that a StringIO instance does. This brings the api as close to
+    compatible as possible with StringIO so that it may be used as a near
+    drop-in replacement to save memory.
+
+    Another issue with SpooledTemporaryFile is that the spooled file is always
+    a cStringIO rather than a StringIO which causes issues with some of our
+    tools.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, max_size=5000000):
+        self._max_size = max_size
+
+    @abstractmethod
+    def read(self, n=-1):
+        """Read n characters from the buffer"""
+        pass
+
+    @abstractmethod
+    def write(self, s):
+        """Write into the buffer"""
+        pass
+
+    @abstractmethod
+    def readline(self, length=None):
+        """Returns the next available line"""
+        pass
+
+    @abstractmethod
+    def readlines(self, sizehint=0):
+        """Returns a list of all lines from the current position forward"""
+        pass
+
+    @abstractmethod
+    def rollover(self):
+        """Roll file-like-object over into a real temporary file"""
+        pass
+
+    @abstractproperty
+    def buffer(self):
+        """Should return a flo instance"""
+        pass
+
+    @abstractproperty
+    def _rolled(self):
+        """Returns whether the file has been rolled to a real file or not"""
+        pass
+
+    def _get_softspace(self):
+        return self.buffer.softspace
+
+    def _set_softspace(self, val):
+        self.buffer.softspace = val
+
+    softspace = property(_get_softspace, _set_softspace)
+
+    @property
+    def _file(self):
+        return self.buffer
+
+    def tell(self):
+        return self._file.tell()
+
+    def seek(self, pos, mode=0):
+        return self.buffer.seek(pos, mode)
+
+    def close(self):
+        return self.buffer.close()
+
+    def flush(self):
+        return self.buffer.flush()
+
+    def isatty(self):
+        return self.buffer.isatty()
+
+    def next(self):
+        return self.readline()
+
+    @property
+    def closed(self):
+        return self.buffer.closed
+
+    @property
+    def pos(self):
+        return self.tell()
+
+    @property
+    def buf(self):
+        return self.getvalue()
+
+    @property
+    def len(self):
+        """Determine the length of the file"""
+        pos = self.tell()
+        if self._rolled:
+            self.seek(0)
+            val = os.fstat(self.fileno()).st_size
+        else:
+            self.seek(0, os.SEEK_END)
+            val = self.tell()
+        self.seek(pos)
+        return val
+
+    def fileno(self):
```

Comment:
@markrwilliams  calling os.fstat(filno_here).st_size behaves differently on pypy. In all other implementations, that call returns the file size. On pypy it returns the number of bytes from the current cursor position within the file. Thus if you don't seek back to zero, st_size reports an incorrect value.